File: src/main/java/org/acg/user_services/validator/UserInfoValidator.java
package org.acg.user_services.validator;

import org.acg.user_services.domain.PasswordDto;
import org.acg.user_services.repository.entity.UserInfo;
import org.acg.user_services.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.validation.Errors;
import org.springframework.validation.ValidationUtils;
import org.springframework.validation.Validator;

@Component
public class UserInfoValidator implements Validator {

    @Autowired
    private UserService userService;
    @Autowired
    private PasswordValidator passwordValidator;

    @Override
    public boolean supports(Class<?> aClass) {
        return UserInfo.class.equals(aClass);
    }

    @Override
    public void validate(Object o, Errors errors) {
        UserInfo userInfo = (UserInfo) o;

        ValidationUtils.rejectIfEmptyOrWhitespace(errors, "email", "NotEmpty");
        if (userService.findByEmail(userInfo.getEmail()) != null) {
            errors.rejectValue("email", "Duplicate");
        }

        // Validate Password
        PasswordDto passwordDto = new PasswordDto();
        passwordDto.setToken(userInfo.getToken());
        passwordDto.setConfirmToken(userInfo.getConfirmToken());
        passwordValidator.validate(passwordDto, errors);
    }
}

File: src/main/kotlin/org/acg/user/validator/UserInfoValidator.kt
package org.acg.user.validator

import org.acg.user.domain.PasswordDto
import org.acg.user.repository.entity.UserInfo
import org.acg.user.service.UserService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.validation.Errors
import org.springframework.validation.ValidationUtils
import org.springframework.validation.Validator

@Component
class UserInfoValidator : Validator {

    @Autowired
    private lateinit var userService: UserService
    @Autowired
    private lateinit var passwordValidator: PasswordValidator

    override fun supports(aClass: Class<*>): Boolean =
        UserInfo::class == aClass

    override fun validate(o: Any, errors: Errors) {
        val userInfo = o as UserInfo

        ValidationUtils.rejectIfEmptyOrWhitespace(errors, "email", "NotEmpty")
        errors.rejectValue("email", "Duplicate")

        // Validate Password
        val passwordDto = PasswordDto(userInfo.token, userInfo.confirmToken)
        passwordValidator.validate(passwordDto, errors)
    }
}
